File: MinimalHeap/MaxHeap.java
package edu.september.designalgorithms;

import java.util.Arrays;

public class MaxHeap {
    private int[] heap;
    private int size;
    private static final int DEFAULT_CAPACITY = 10;

    public MaxHeap() {
        heap = new int[DEFAULT_CAPACITY];
        size = 0;
    }

    /**
     * Метод добавления в конец кучи. Метод вполне стандартный, только в конце происходит поднятие наверх в случае
     * нарушения условий кучи.
     * @see MaxHeap#heapifyUp()
     * @param value элемент целочисленный
     */
    public void add(int value) {
        ensureCapacity();
        heap[size] = value;
        size++;
        heapifyUp();
    }

    /**
     * Получение максимума кучи с последующим удалением его.
     * @return целочисленный максимум
     */
    public int poll() {
        if (size == 0) { throw new IllegalStateException("Heap is empty"); }
        int max = heap[0];
        heap[0] = heap[size-1];
        size--;
        heapifyDown();
        return max;
    }

    /**
     * Смотрим на максимум кучи.
     * @return целочисленное значение максимума
     */
    public int peek() {
        if (size == 0) { throw new IllegalStateException("Heap is empty"); }
        return heap[0];
    }

    /**
     * Метод, который поднимает наверх элементы, которые нарушают условия кучи.
     */
    private void heapifyUp() {
        int index = size - 1;
        while (hasParent(index) && heap[index] > heap[getParentIndex(index)]) {
            swap(index, getParentIndex(index));
            index = getParentIndex(index);
        }
    }

    /**
     * Нормализация кучи после удаления корня.
     */
    private void heapifyDown() {
        int index = 0;
        while (hasLeftChild(index)) {
            int largerChildIndex = getLeftChildIndex(index);
            if (hasRightChild(index) &&
            heap[getRightChildIndex(index)] > heap[largerChildIndex]) {
                largerChildIndex = getRightChildIndex(index);
            }

            if (heap[index] >= heap[largerChildIndex]) { break; }

            swap(index, largerChildIndex);
            index = largerChildIndex;
        }
    }

    /*
    Методы-геттеры для получения индекса левого/правого ребёнка и родителя.
     */
    private int getLeftChildIndex(int parent) { return 2 * parent + 1; }
    private int getRightChildIndex(int parent) { return 2 * parent + 2; }
    private int getParentIndex(int child) { return (child - 1) / 2; }

    private boolean hasLeftChild(int index) { return getLeftChildIndex(index) < size;}
    private boolean hasRightChild(int index) { return getRightChildIndex(index) < size;}
    private boolean hasParent(int index) { return index > 0; }

    /**
     * Сваппер
     * @param i индекс свапаемого элемента
     * @param j индекс свапающего элемента
     */
    private void swap(int i, int j) {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    /**
     * Метод, увеличивающий кучу, если мы превысим размер текущий
     * (начиная с {@code DEFAULT_CAPACITY})
     */
    private void ensureCapacity() {
        if (size == heap.length) {
            heap = Arrays.copyOf(heap, heap.length * 2);
        }
    }

    /**
     * Вывод кучи.
     */
    public void viewHeap() {
        for (int i = 0; i < size; ++i) {
            System.out.print(heap[i] + " ");
        }
        System.out.println();
    }
}

File: Contest №1/MinimalHeap/MinimalHeap/MaxHeap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MinimalHeap
{
    internal class MaxHeap
    {
        private int[] heap;
        private int size;
        private const int DEFAULT_CAPACITY = 10;

        public MaxHeap()
        {
            heap = new int[DEFAULT_CAPACITY];
            size = 0;
        }

        public int Peek()
        {
            if (size == 0)
            {
                throw new Exception("Heap is empty.");
            }
            return heap[0];
        }

        public void Add(int val)
        {
            EnsureCapacity();
            heap[size++] = val;
            HeapifyUp();
        }

        public void Add(int[] numArray)
        {
            foreach (int num in numArray)
            {
                this.Add(num);
            }
        }

        private void HeapifyUp()
        {
            int index = size - 1; // last element.

            while (HasParent(index))
            {
                int parentIndex = GetParentIndex(index);
                if (heap[parentIndex] >= heap[index]) { break; }
                Swap(index, parentIndex);
                index = parentIndex;
            }
        }

        public int Poll()
        {
            if (size == 0) { throw new Exception("Heap is empty!"); }
            int max = heap[0];
            heap[0] = heap[size - 1];
            size--;
            HeapifyDown();
            return max;
        }

        private void HeapifyDown()
        {
            int index = 0;
            while (HasLeftChild(index))
            {
                int smallerLeftChild = GetLeftChildIndex(index);

                if (HasRightChild(index)
                    && heap[GetLeftChildIndex(index)] < heap[GetRightChildIndex(index)])
                {
                    smallerLeftChild = GetRightChildIndex(index);
                }

                if (heap[index] >= heap[smallerLeftChild])
                {
                    break;
                }

                Swap(index, smallerLeftChild);
                index = smallerLeftChild;
            }
        }

        public int[] HeapSort()
        {
            int originalSize = size;
            int[] sortedHeap = new int[size];

            for (int i = 0; i < sortedHeap.Length; i++)
            {
                sortedHeap[i] = Poll();
            }

            heap = new int[DEFAULT_CAPACITY];
            size = 0;

            Add(sortedHeap);
            return sortedHeap;
        }

        // Getters.
        private int GetLeftChildIndex(int parent) { return 2 * parent + 1; }
        private int GetRightChildIndex(int parent) { return 2 * parent + 2; }
        private int GetParentIndex(int child) { return (child - 1) / 2; }

        // Checkers
        private bool HasLeftChild(int parent) { return GetLeftChildIndex(parent) < size; }
        private bool HasRightChild(int parent) { return GetRightChildIndex(parent) < size; }
        private bool HasParent(int child) { return child > 0; }

        private void Swap(int first, int second)
        {
            int temp = heap[first];
            heap[first] = heap[second];
            heap[second] = temp;
        }
        private void EnsureCapacity()
        {
            if (size == heap.Length)
            {
                int[] increasedHeap = new int[heap.Length * 2];
                Array.Copy(heap, increasedHeap, heap.Length);
                heap = increasedHeap;
            }
        }

        public void PrintHeap()
        {
            for (int i = 0; i < size; ++i)
            {
                Console.Write($"{heap[i]} ");
            }
            Console.WriteLine();
        }
    }
    internal class Program
    {
        static void Main(string[] args)
        {
            int commandsValue = int.Parse(Console.ReadLine());
            MaxHeap myHeap = new MaxHeap();
            List<int> printMaxis = new List<int>();

            for (int cmd = 0; cmd < commandsValue; ++cmd)
            {
                string line = Console.ReadLine();
                string[] tokens = line.Split();

                if (tokens[0] == "0")
                {
                    myHeap.Add(int.Parse(tokens[1]));
                }
                else
                {
                    printMaxis.Add(myHeap.Poll());
                }
            }

            foreach (int maxs in printMaxis)
            {
                Console.WriteLine(maxs);
            }
        }
    }
}